src/cart/crt_iv.c

```diff
@@ -2867,7 +2869,8 @@ bulk_update_transfer_done_aux(const struct crt_bulk_cb_info *info)
 
 			D_GOTO(exit, rc);
 		}
-		output->rc = -DER_SUCCESS;
+		rc = crt_bulk_free(cb_info->buc_bulk_hdl);
+		output->rc = rc;
```

Comment:
iv_on_put() and bulk_free needs to switch order indeed.  

I do not see yield between these calls, so memory may not be able to be ruined between these calls? or you see potential yield somewhere? 

L2888 and 2891 needs to be fixed as well?

Reply:
crt_bulk_free() should prevent remote peer from accessing the buffer so it should not matter what happens after bulk free returns.

and indeed wrt lines 2888-2891

Reply:
>crt_bulk_free() should prevent remote peer from accessing the buffer so it should not matter what happens after bulk free returns.

I mean if progress won't be called after free() and before bulk_free(), is it possible the registered memory being ruined between these two call?

Reply:
indeed, thanks

Reply:
I think provider may have progress thread, and RDMA capable network can even do this w/o progress thread

Reply:
@wangdi1 well, you are doing crt_bulk_free() asynchronously with possible rdma remote transfer, so its possible the memory will contain partial data or some garbage, but the buffer contents should not matter anymore as on_put() will be done next to free the buffer

Reply:
> I think provider may have progress thread, and RDMA capable network can even do this w/o progress thread

indeed.

Reply:
> @wangdi1 well, you are doing crt_bulk_free() asynchronously with possible rdma remote transfer, so its possible the memory will contain partial data or some garbage, but the buffer contents should not matter anymore as on_put() will be done next to free the buffer

yeah, free() should be always done after bulk_free(). Though I wonder if in this case the provider can ruined the register memory(freed) without calling progress, note: this is in VM environment.

Reply:
Is it possible that freed memory being immediately allocated by other thread/xstream before calling the crt_bulk_free()? Then the rdma transfer could corrupt the allocated memory.

Reply:
I am not sure I understand @NiuYawei , this change specifically makes freeing of memory (ivo_on_put()) to be performed after crt_bulk_free, so until ivo_on_put() that memory will not be handed out to anyone else

Reply:
> I am not sure I understand @NiuYawei , this change specifically makes freeing of memory (ivo_on_put()) to be performed after crt_bulk_free, so until ivo_on_put() that memory will not be handed out to anyone else

right, I mean before this fix, the ivo_on_put() will free memory before the crt_bulk_free(), and the freed memory could be reused by other xstream/thread and being corrupted by rdma transfer. I was replying Di's question of why it could cause corruption even there is no yield between iov_on_put() and crt_bulk_free().

Reply:
ah, ok sorry:)

Reply:
Looks few other places are overlooked? like in handle_ivfetch_response(), crt_bulk_free() is called after the ivo_on_refresh() callback, in handle_ivupdate_response(), crt_bulk_free() could be called after ivo_on_put() in error case.

Reply:
indeed, let me fix this too

Reply:
> Looks few other places are overlooked? like in handle_ivfetch_response(), crt_bulk_free() is called after the ivo_on_refresh() callback, in handle_ivupdate_response(), crt_bulk_free() could be called after ivo_on_put() in error case.

I think ivo_on_refresh() is ok right? it's not going to free buffer.

Reply:
ivo_on_refresh() and all other ivo_* should be ok other than ivo_on_put() which should be the only one freeing the memory